[PateShin] iP

PateShin opened the pull request:
# FICIN

FICIN is ~~easy~~ **VERY EASY** chat bot which can help you to track and record your tasks by...

- **TODO**
- **DEADLINE** 
- **EVENT**

**TODO**: Simply record your task. Here is an example:
```
____________________________________________________________
Todo Study // <- This is the command "TODO" and your task
____________________________________________________________
```

**DEADLINE**: Set deadline for your task with command ```/by```.
```
____________________________________________________________
Deadline Assignment /by tomorrow // <- This is the command "Deadline" and your task
____________________________________________________________
```

**EVENT**: Set event time for your task with command ```/from ... /to ...```.
```
____________________________________________________________
Event Travel /from today /to 2024/04/29 // <- This is the command "Event" and your task
____________________________________________________________
```

Furthermore, you are able to **Mark** & **Unmark** tasks, **Delete** tasks and **Find** tasks! 😮 
These can be done by following:

```
- List // To see the list of tasks
- Delete 1 // Number of task in list
- Mark 1
- Unmark 1
- Find // TYPE YOUR TASK TO FIND 🔤
```

Features: 

- [x] Loading & Saving Task 
- [x] Tasks - TODO, DEADLINE, EVENT)
- [x] Date Command for Deadline and Event (TODAY, TOMORROW)
- [x] Commands - DELETE, MARK, UNMARK, FIND
- [x] Create GUI (Coming soon) 

Here is the [link](https://github.com/PateShin/ip) to Github.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

PateShin committed:
Add features

PateShin committed:
Change print out

PateShin committed:
A-TextUiTesting

PateShin committed:
Abstract out Duke

PateShin committed:
Fix file not existing error

PateShin committed:
Add today and tomorrow to for deadline and tomorrow

PateShin committed:
Make package

PateShin committed:
Merge remote-tracking branch 'origin/add-gradle-support'

PateShin committed:
Make jar file

PateShin committed:
Add javadoc

PateShin committed:
Fix functions names

PateShin committed:
Remove comment

PateShin committed:
Fix line spacing

PateShin committed:
Add javadoc to Duke.java

PateShin committed:
Add find command

PateShin committed:
Add no filtered task found

PateShin committed:
Delete unnecessary comments

PateShin committed:
Merge branch 'branch-Level-9' into branch-A-JavaDoc

PateShin committed:
Add javadoc to FindCommand and Ui

PateShin committed:
Rewrite deadline code structure

PateShin committed:
Merge branch 'branch-A-JavaDoc' into branch-A-CodingStandard

# Conflicts:
#	src/main/java/duke/CommandType.java

PateShin committed:
Rewrite code structure

PateShin committed:
Fix load tasks

PateShin committed:
Fix code style

PateShin committed:
Fix code style

PateShin committed:
Create GUI

PateShin committed:
Fix on message

jjchee77 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,197 @@
+                return new DeleteCommand();
+            case BYE:
+                return new ExitCommand();
+            case TODO:
+            case DEADLINE:
+            case EVENT:
```
Could perhaps add "// Fallthrough" for cases without the break or return keyword.
https://se-education.org/guides/conventions/java/intermediate.html

PateShin replied:
// Fallthrough has added

jjchee77 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * Unmarks the task as done by setting the completion status to false.
+     */
+    public void unmarkAsDone() {
```
"unmarkAsDone" might sound a little weird as it may be interpreted as the final state of the task is "done". "markAsNotDone" could be a better alternative.

PateShin replied:
I like the alternative that you suggest. The "unmarkAsDone" has changed to "markAsNotDone".

jjchee77 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,82 @@
+     *
+     * @return A string representing the task's completion status.
+     */
+    public String getStatusNumber() {
```
Method name implies that the return type is a number but actual return type is a string.

PateShin replied:
Thanks for pointing out. I have now changed the String Type to Integer Type.

jjchee77 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,99 @@
+     * @param index The index of the task to be unmarked, 1-based.
+     * @return The task that was unmarked, or null if the index is invalid.
+     */
+    public Task unmarkTask(int index) {
```
Perhaps could incorporate the word "done" here for consistency among other parts of the code.

PateShin replied:
"unmarkTask" has now changed to "unmarkTaskDone"

jjchee77 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,99 @@
+     * @param index The index to be validated, 1-based.
+     * @return True if the index is valid, false otherwise.
+     */
+    private boolean isValidIndex(int index) {
```
Argument `index` could possibly be renamed to `taskNum` as it does not represent the actual index of the task within the list. Similar occurrences in other parts of the code.

PateShin replied:
That is actually interesting recommendation! I am taking your opinion, so now ```index``` is renamed to ```taskNum```.

jjchee77 commented on src/main/java/duke/task/Todo.java:
```diff
@@ -0,0 +1,38 @@
+     * @return A formatted string representing the Todo object.
+     */
+    @Override
+    public String toFileString() {
```
"toFileString" may be a little difficult to interpret the method's behaviour at a glance. "toSaveableString/Format" could be a better alternative.

PateShin replied:
Personally, ```toFileString``` is still acceptable. However, ```toSaveableFormat``` has better to understand what it is for. Therefore, ```toFileString``` is changed to ```toSaveableFormat```.

jjchee77 reviewed (commented):
Code mostly conforms to the **Code Quality: Naming** standards :) Good job!
Maybe a few areas that could be improved

PateShin committed:
Design GUI

PateShin committed:
Design GUI

PateShin committed:
Merge branch 'branch-Level-10' into branch-A-CheckStyle

PateShin committed:
Check code style

PateShin committed:
Fix varaibles names

kailinteoo commented on src/main/java/duke/DeleteCommand.java:
```diff
@@ -0,0 +1,37 @@
+     * @throws DukeException If there is an issue executing the command or if the task index is invalid.
+     */
+    @Override
+    public void execute(TaskList taskList, Ui ui, String userInput) throws DukeException {
+        try {
+            int taskIndex = Parser.extractTaskIndex(userInput);
+            if (taskIndex <= 0 || taskIndex > taskList.getSize()) {
+                throw new DukeException("Invalid task number. Please try again.");
+            }
+            Task removedTask = taskList.deleteTask(taskIndex);
+            ui.showTaskDeleted(removedTask, taskList.getSize());
+        } catch (NumberFormatException e) {
+            throw new DukeException("Invalid number format. Please enter a valid task number.");
+        } catch (Exception e) {
+            throw new DukeException("An error occurred while trying to delete the task: " + e.getMessage());
```
I like the clarity and simplicity in the way you name your variables. It makes it easy for users to understand your code without having to refer to other classes.

kailinteoo commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,12 @@
+import javafx.application.Application;
+
+/**
+ * A launcher class to workaround classpath issues.
+ */
+public class Launcher {
+    public static void main(String[] args) {
+        Application.launch(Main.class, args);
+    }
+}
```
Maybe providing further clarification on the specific classpath issues you mentioned in your JavaDoc would help to enhance understanding for future developers or reader reviewing your code.

PateShin replied:
I have provided the clarification on the specific class path issues so other developers can review my code with improved readability.

kailinteoo commented on src/main/java/duke/ListCommand.java:
```diff
@@ -0,0 +1,30 @@
+     *                  It is not used here, but required by the method signature).
+     */
+    @Override
+    public void execute(TaskList taskList, Ui ui, String userInput) {
+        if (taskList.getSize() == 0) {
+            ui.showNoTask();
+        } else {
+            ui.showTasks(taskList.getTasks());
```
Your code arrangement is commendable, as it is clear and straightforward for users to understand with just a single glance of it.

kailinteoo commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,199 @@
+        try {
+            CommandType command = Parser.parseCommand(userInput);
+
+            switch (command) {
+            case LIST:
+                return new ListCommand();
+            case MARK:
+                return new MarkCommand();
+            case UNMARK:
+                return new UnmarkCommand();
+            case DELETE:
+                return new DeleteCommand();
+            case BYE:
+                return new ExitCommand();
+            case TODO:
+                // Fallthrough
+            case DEADLINE:
+                // Fallthrough
+            case EVENT:
+                return new AddCommand();
+            case FIND:
+                return new FindCommand();
+            default:
+                throw new DukeException("I'm sorry, but I don't know what that means :-(");
+            }
+        } catch (IllegalArgumentException e) {
+            throw new DukeException("I'm sorry, but I don't know what that means :-(");
```
I admire your implementation of the switch statement in this section of your code to determine whether the user has entered the correct input. Personally felt that it was a clever approach that I had not considered before.

kailinteoo commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,154 @@
+     * @param line The line of text representing a task.
+     * @return The Task object created from the line.
+     */
+    private static Task createTaskFromLine(String line) {
```
Maybe can add more explanatory comments to clarify the purpose of each variables used and their roles would certainly enhance understanding, especially for uses who may not be familiar with the codebase yet.

PateShin replied:
I have added detailed comments to do the method.

kailinteoo reviewed (commented):
In general, I am quite impressed with the structure of your code and your implementation of OOP principles. It offers a seamless and easily comprehensible experience for users who are reviewing your code for the first time, allowing them to grasp the functionality of each class and the relationships between them. Your Javadoc is also well-written, although I would suggest adding a bit more detail to certain classes I have commented on, as it would further aid understanding of individual functions and variables. Nevertheless, overall, GOOD JOB!

PateShin committed:
Improve JavaDoc in Storage and Launcher classes

More detailed JavaDoc is required in the Launcher class and the
createTaskFromLine method in the Storage class for better clarity.

Let's
* Added explanations in Storage class to make it clear how tasks are
saved and loaded. This helps new developers understand the file
handling part quickly.
* Improved Launcher class comments to explain why it's used, focusing
on solving classpath issues with JavaFX applications.

These updates aim to make the code easier to understand, helping
developers get up to speed faster.

PateShin committed:
Add assertions to Duke for better error checking

The chatbot did not check assumptions in its code, which could lead to unexpected bugs.

Let's,
* check taskList is not empty after loading tasks
* check that user inputs are not empty before processing
* ensure ui and taskList are ready before responding to commands
* verify command outputs are not empty

These checks help find problems early, making the app more reliable.

PateShin committed:
Refactor classes to enhance code quality

PateShin committed:
Merge pull request #2 from PateShin/branch-A-Assertions

Add assertions to Duke for better error checking

PateShin committed:
Merge branch 'master' into branch-A-CodeQuality

PateShin committed:
Merge pull request #1 from PateShin/branch-A-CodeQuality

PR for Code Quality

PateShin committed:
Refactor FindCommand to Use Java Streams

This commit updates the execute() method in the FindCommand class to
utilize Java Streams for task filtering. Using Streams makes the code
more expressive and succinct, which enhances readability and
maintainability without changing the way the class behaves.

PateShin committed:
Add help command

Implemented a new command help that shows for first user, and provides
users with a list of available commands.

* Added a showHelpMessage() method to the Ui class to display the list
of commands.
* Added HELP command to Parser and CommandType
* Added HelpCommand class
* Modified the handleUserInput() method in the MainWindow class to
handle the help command input.
* Updated the initialize method in the MainWindow class to display the
help message as the initial response when the application starts.

This changes improve user experience by offering  access to information
about available commands, helping users utilize the application
more effectively.

PateShin committed:
Fix code Style for Ui and MainWindow classes

PateShin committed:
Add yml file to the new directory .github/workflows

PateShin committed:
Enhance DialogBox and Help Message

This commits introduce the changes in Ui help messages and DialogBox

Changes include:
- The DialogBox is now show full message without any cut
- The help message is imporved for both MainWindow and Ui

These enhancements aim to improve user experience

PateShin committed:
Enhance GUI Personality and Messaging

This commit introduces several adjustments to images, colors, and
response messages to enhance the personality.

Key changes include:
- Updated messages in the UI class to add uniqueness to the GUI.
- Personalized color scheme to enhance visual appeal.
- Added images to infuse more personality into the interface.

These enhancements aim to improve the uniqueness of the GUI.

PateShin committed:
Merge branch 'branch-A-BetterGui' into branch-A-Personality

PateShin committed:
Improve Error Handling in Parser

This commit enhances the Parser with additional error handling.

Key changes include:
- Added an error message to handle empty input.

This improvement aim to enhance the user experience.

PateShin committed:
Merge branch 'branch-A-Personality'

PateShin committed:
Add JUnit Tests for Todo Class

PateShin committed:
Add User Guide

This commit introduces the addition of User Guide

Key Changes
- Explanation of user command has included

This implementation will make users to easy to use the Chat bot

PateShin committed:
Improve Javadoc in Duke.java

PateShin committed:
Changes for gradle file

This commits describes the change in build.gradle

Key Change:
* Gradle is now creating jar file that runs GUI appliction

PateShin committed:
Fix minor bug

PateShin committed:
Remove TodoTest

PateShin committed:
Fix minor bugs

PateShin committed:
Fix runtest.bat and .sh file

PateShin committed:
Fix minor issues

PateShin committed:
Fix runtest files and expected output

PateShin committed:
Fix runtest.sh file and Expected.txt

PateShin committed:
Fix Expected.txt

PateShin committed:
Fix Ui.java message

Key Change:
* Space between showGoodBye and Help message has changed

PateShin committed:
Fix the format of TodoTest

PateShin committed:
Undo the change
